internal/rpc/digest_md5.go

```diff
@@ -0,0 +1,426 @@
+
+	ourToken := &hadoop_common.TokenProto{}
+	ourToken.Kind = token.Kind
+	ourToken.Password = token.Password[:]
+	ourToken.Service = token.Service
+	ourToken.Identifier = token.GetIdentifier()
+
+	// if the server didn't send us a Nonce, then the data isn't encrypted
+	// but we will still attempt to authenticate
+	if useSecure && len(key.Nonce) == 0 {
+		useSecure = false
+	}
+
+	if useSecure {
+		nonceBase64Len := int(math.Ceil(4 * (float64(len(key.Nonce)) / 3)))
+		buf := new(bytes.Buffer)
+		buf.Grow(6 + nonceBase64Len + len(key.BlockPoolID))
+		buf.WriteString(fmt.Sprintf("%d", key.KeyID))
+		buf.WriteString(" ")
+		buf.WriteString(key.BlockPoolID)
+		buf.WriteString(" ")
+		buf.WriteString(base64.StdEncoding.EncodeToString(key.Nonce))
+		ourToken.Identifier = buf.Bytes()
+		ourToken.Password = key.EncryptionKey
+	} else {
+		ourToken.Identifier = make([]byte, base64.StdEncoding.EncodedLen(len(token.GetIdentifier())))
+		base64.StdEncoding.Encode(ourToken.Identifier, token.GetIdentifier())
+	}
+
+	dgst := digestMD5{
+		authID:   ourToken.Identifier,
+		passwd:   base64.StdEncoding.EncodeToString(ourToken.Password),
+		hostname: auth.GetServerId(),
+		service:  auth.GetProtocol(),
+	}
+
+	// start handshake
+	var err error
+	if err = binary.Write(conn, binary.BigEndian, uint32(0xDEADBEEF)); err != nil {
+		return nil, err
+	}
+
+	if err = sendSaslMsg(conn, hdfs.DataTransferEncryptorMessageProto_SUCCESS.Enum(), []byte{}, "", false); err != nil {
+		return nil, err
+	}
+
+	msg := &hdfs.DataTransferEncryptorMessageProto{}
+	if err = readPrefixedMessage(conn, msg); err != nil {
+		return nil, err
+	}
+
+	challengeResponse, err := dgst.challengeStep1(msg.Payload)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = sendSaslMsg(conn, hdfs.DataTransferEncryptorMessageProto_SUCCESS.Enum(), []byte(challengeResponse), "", useSecure); err != nil {
+		return nil, err
+	}
+
+	if err = readPrefixedMessage(conn, msg); err != nil {
+		return nil, err
+	}
+
+	// the result of this challenge evaluation should be an empty string
+	if challengeResponse, err = dgst.challengeStep2(msg.Payload); err != nil || len(challengeResponse) != 0 {
+		return nil, err
+	}
+
+	// we're only authenticating, no RPC protection involved, so no wrapping the connection
+	// just return the connection as is, we're done here!
+	if dgst.token.qop == qopAuthentication {
+		return conn, nil
+	}
+
+	kic, kis := generateIntegrityKeyPair(dgst.a1())
+
+	var d digestMD5Handler
+	if dgst.token.qop == qopPrivacy {
+		if dgst.cipher == cipherUnknown {
+			return nil, fmt.Errorf("Could not find implemented cipher among choices: %v", dgst.token.cipher)
+		}
+		kcc, kcs := generatePrivacyKeyPair(dgst.a1(), dgst.cipher)
+		d, err = newDigestMD5PrivacyConn(conn, kic[:], kis[:], kcc[:], kcs[:])
+		if err != nil {
+			return nil, err
+		}
+	} else if dgst.token.qop == qopIntegrity {
+		d = newDigestMD5IntegrityConn(conn, kic[:], kis[:])
+	}
+
+	if len(msg.GetCipherOption()) > 0 {
```

Comment:
Shouldn't this always be true if the qop is `privacy`? Put another way, is `digestMD5PrivacyConn` ever actually used (beyond the call to `decode` below)?

Reply:
if for some reason the server doesn't support AES then this would be empty and it would only use the rc4 that was negotiated during the digest authentication.